File: backend/app/schemas/stock_movement.py
from datetime import date, datetime
from typing import Literal

from pydantic import BaseModel

StockMovementReasonLiteral = Literal[
    "manual_edit",
    "add_stock",
    "purchase_order_received",
    "purchase_order_receive_undone",
    "order_consumed",
    "order_dispatch_undone",
    "assembly_build",
    "assembly_build_material",
    "manual_item_consumed",
]

ComponentTypeLiteral = Literal["part", "assembly"]


class StockMovementRead(BaseModel):
    id: int
    component_type: ComponentTypeLiteral
    component_id: int
    component_name: str
    quantity: int
    reason: StockMovementReasonLiteral
    note: str | None
    order_id: int | None
    order_unique_key: str | None = None
    purchase_order_id: int | None
    shop_id: int | None
    created_at: datetime


class StockMovementListRead(BaseModel):
    items: list[StockMovementRead]
    total: int


class ConsumptionSummaryPoint(BaseModel):
    date: date
    quantity: int


class ConsumptionSummarySeries(BaseModel):
    component_type: ComponentTypeLiteral
    component_id: int
    component_name: str
    total: int
    points: list[ConsumptionSummaryPoint]


class ConsumptionSummaryRead(BaseModel):
    days: int
    series: list[ConsumptionSummarySeries]


class AssemblyBuildSummaryPoint(BaseModel):
    date: date
    quantity: int


class AssemblyBuildSummaryRead(BaseModel):
    days: int
    points: list[AssemblyBuildSummaryPoint]

File: frontend/src/types/stock_movement.ts
export type StockMovementReason =
  | "manual_edit"
  | "add_stock"
  | "purchase_order_received"
  | "purchase_order_receive_undone"
  | "order_consumed"
  | "order_dispatch_undone"
  | "assembly_build"
  | "assembly_build_material"
  | "manual_item_consumed";

export type StockMovementComponentType = "part" | "assembly";

export interface StockMovement {
  id: number;
  component_type: StockMovementComponentType;
  component_id: number;
  component_name: string;
  quantity: number;
  reason: StockMovementReason;
  note: string | null;
  order_id: number | null;
  order_unique_key: string | null;
  purchase_order_id: number | null;
  shop_id: number | null;
  created_at: string;
}

export interface StockMovementListResult {
  items: StockMovement[];
  total: number;
}

export interface StockMovementFilters {
  component_type?: StockMovementComponentType;
  part_id?: number;
  assembly_id?: number;
  reason?: StockMovementReason;
  order_unique_key?: string;
  date_from?: string;
  date_to?: string;
  shop_id?: number;
  include_shared?: boolean;
}

export interface ConsumptionSummaryPoint {
  date: string;
  quantity: number;
}

export interface ConsumptionSummarySeries {
  component_type: StockMovementComponentType;
  component_id: number;
  component_name: string;
  total: number;
  points: ConsumptionSummaryPoint[];
}

export interface ConsumptionSummary {
  days: number;
  series: ConsumptionSummarySeries[];
}

export interface AssemblyBuildSummaryPoint {
  date: string;
  quantity: number;
}

export interface AssemblyBuildSummary {
  days: number;
  points: AssemblyBuildSummaryPoint[];
}
